File: CPP/Film.cpp
#include <iostream>
#include <string>
#include <iomanip> 

using namespace std;

//deklarasi class Film untuk merepresentasikan data film di Bioskop
class Film
{
    private:
        //private atribut, hanya bisa diakses dari dalam class ini sendiri 
        string id_film;      //ID unik untuk tiap film
        string judul_film;   //judul film yang tayang
        string sutradara;    //nama sutradara film
        int stok_tiket;      //jumlah tiket/kursi yang masih tersedia
        double harga_tiket;  //harga tiket untuk menonton film ini

    public:
    // constructor, dipanggil otomatis saat objek Film dibuat
    Film(string id, string judul, string sutradara, int stok, double harga)
    {
        setId(id);           //inisialisasi id lewat setter
        setJudul(judul);     //inisialisasi judul lewat setter
        setSutradara(sutradara); //inisialisasi sutradara lewat setter
        setStok(stok);       //inisialisasi stok lewat setter 
        setHarga(harga);     //inisialisasi harga lewat setter 
    }

    //setter untuk merubah value id
    void setId(const string& id)
    {
        this->id_film = id; //inisialisasi/ubah atribut id_film
    }

    //setter untuk merubah value judul
    void setJudul(const string& judul)
    {
        this->judul_film = judul; //inisialisasi/ubah atribut judul_film
    }

    //setter untuk merubah value sutradara
    void setSutradara(const string& sutradara)
    {
        this->sutradara = sutradara; //inisialisasi/ubah atribut sutradara
    }

    //setter untuk merubah value stok, dengan validasi
    void setStok(const int& stok)
    {
        if (stok >= 0) //stok tidak boleh negatif
        {
            this->stok_tiket = stok; //inisialisasi/ubah atribut stok_tiket
        }
        else
        {
            cout << "Stok tiket tidak boleh negatif." << endl; //pesan jika input tidak valid
        }
    }

    //setter untuk merubah value harga, dengan validasi
    void setHarga(const double& harga)
    {
        if (harga > 0) //harga harus lebih dari 0
        {
            this->harga_tiket = harga; //inisialisasi/ubah atribut harga_tiket
        }
        else
        {
            cout << "Harga tiket harus lebih dari 0." << endl; //pesan jika input tidak valid
        }
    }

    // getter untuk mengambil value id
    string getId() const
    {
        return id_film; //mengembalikan nilai id_film
    }

    // getter untuk mengambil value judul
    string getJudul() const
    {
        return judul_film; //mengembalikan nilai judul_film
    }

    // getter untuk mengambil value sutradara
    string getSutradara() const
    {
        return sutradara; //mengembalikan nilai sutradara
    }

    // getter untuk mengambil value stok
    int getStok() const
    {
        return stok_tiket; //mengembalikan nilai stok_tiket
    }

    // getter untuk mengambil value harga
    double getHarga() const
    {
        return harga_tiket; //mengembalikan nilai harga_tiket
    }

    //prosedur untuk menampilkan seluruh data film
    void tampilkanData() const
    {
        //print seluruh atribut film ke layar
        cout << "ID           : " << getId() << endl
            << "Judul Film   : " << getJudul() << endl
            << "Sutradara    : " << getSutradara() << endl
            << "Stok Tiket   : " << getStok() << endl
            << "Harga Tiket  : Rp" << fixed << setprecision(0) << getHarga() << endl;
    }

    //destructor, dipanggil otomatis saat objek dihapus/keluar dari scope
    ~Film()
    {
    }
};

File: Python/Film.py
class Film:
    # constructor, dipanggil otomatis saat objek Film dibuat
    def __init__(self, id_film: str, judul_film: str, sutradara: str, stok_tiket: int, harga_tiket: float):

        self.__id_film = str(id_film)         #inisialisasi atribut id_film (private, encapsulation)
        self.__judul_film = str(judul_film)   #inisialisasi atribut judul_film
        self.__sutradara = str(sutradara)     #inisialisasi atribut sutradara
        self.__stok_tiket = int(stok_tiket)   #inisialisasi atribut stok_tiket
        self.__harga_tiket = float(harga_tiket) #inisialisasi atribut harga_tiket

    # Getter untuk mengambil data
    def getId(self):
        return self.__id_film #mengembalikan nilai id_film

    def getJudul(self):
        return self.__judul_film #mengembalikan nilai judul_film

    def getSutradara(self):
        return self.__sutradara #mengembalikan nilai sutradara

    def getStok(self):
        return self.__stok_tiket #mengembalikan nilai stok_tiket

    def getHarga(self):
        return self.__harga_tiket #mengembalikan nilai harga_tiket

    # Setter untuk merubah value
    def setId(self, id_film):
        self.__id_film = id_film #mengubah value atribut dengan value baru

    def setJudul(self, judul_film):
        self.__judul_film = judul_film #mengubah value atribut dengan value baru

    def setSutradara(self, sutradara):
        self.__sutradara = sutradara #mengubah value atribut dengan value baru

    def setStok(self, stok_tiket):
        if stok_tiket >= 0: #tidak boleh angka negatif
            self.__stok_tiket = stok_tiket #mengubah value atribut dengan value baru
        else:
            print("Stok tiket tidak boleh negatif.") #pesan jika angka negatif

    def setHarga(self, harga_tiket):
        if harga_tiket > 0: #tidak boleh angka negatif/nol
            self.__harga_tiket = harga_tiket #mengubah value atribut dengan value baru
        else:
            print("Harga tiket harus lebih dari 0.") #pesan jika angka tidak valid

    #prosedur untuk menampilkan seluruh data film
    def tampilkanData(self):
        #print seluruh atribut ke layar
        print("ID           :", self.getId())
        print("Judul Film   :", self.getJudul())
        print("Sutradara    :", self.getSutradara())
        print("Stok Tiket   :", self.getStok())
        print("Harga Tiket  : Rp", self.getHarga())
